shadowsocks/stream.go

```diff
@@ -80,7 +82,8 @@ func (sw *shadowsocksWriter) encryptBlock(ciphertext []byte, plaintext []byte) (
 }
 
 func (sw *shadowsocksWriter) Write(p []byte) (int, error) {
-	n, err := sw.ReadFrom(bytes.NewReader(p))
+	sw.input.Reset(p)
```

Comment:
Can you create the Reader here instead? Shouldn't this put the Reader on the stack with zero allocations?

```go
func (sw *shadowsocksWriter) Write(p []byte) (int, error) {
    Reader reader
    reader.Reset(p)
    n, err := sw.ReadFrom(bytes.NewReader(p))
    return int(n), err
}
```

Reply:
I tried this:
```go
var input bytes.Reader
input.Reset(p)
n, err := sw.ReadFrom(&input)
```
It didn't work; the allocation was not avoided.

`go build -gcflags="-m=2" .` shows the escape analysis.  It says that ReadFrom's argument "escapes to the heap" because `ReadFrom(r)` calls `r.Read(...)`, which could do anything (e.g store `r` in a global variable).